webapp/lib/controller.dart

```diff
@@ -690,16 +778,45 @@ void setMessageTag(model.Tag tag, model.Message message, model.Conversation conv
   }
 }
 
-Set<model.Conversation> filterConversationsByTags(Set<model.Conversation> conversations, List<model.Tag> filterTags, DateTime afterDateFilter) {
-  if (filterTags.isEmpty && afterDateFilter == null) return conversations;
+Set<model.Conversation> filterConversationsByTags(Set<model.Conversation> conversations, ConversationFilter conversationFilter) {
+  if (conversationFilter.includeTags.isEmpty &&
+      conversationFilter.excludeTags.isEmpty &&
+      conversationFilter.includeAfterDateFilter == null &&
+      conversationFilter.excludeAfterDateFilter == null) {
+    return conversations;
+  }
 
   var filteredConversations = emptyConversationsSet;
-  var filterTagIds = filterTags.map<String>((tag) => tag.tagId).toList();
+  var includeFilterTagIds = conversationFilter.includeTags.map<String>((tag) => tag.tagId).toSet();
+  var excludeFilterTagIds = conversationFilter.excludeTags.map<String>((tag) => tag.tagId).toSet();
   conversations.forEach((conversation) {
-    // Filter by the last (most recent) conversation
-    // TODO consider an option to filter by the first conversation
-    if (afterDateFilter != null && conversation.messages.last.datetime.isBefore(afterDateFilter)) return;
-    if (!conversation.tagIds.toSet().containsAll(filterTagIds)) return;
+    // Filter by the last (most recent) message
+    // TODO consider an option to filter by the first message
+    if (conversationFilter.includeAfterDateFilter != null && conversation.messages.last.datetime.isBefore(conversationFilter.includeAfterDateFilter)) return;
+    if (conversationFilter.excludeAfterDateFilter != null && conversation.messages.last.datetime.isAfter(conversationFilter.excludeAfterDateFilter)) return;
+
+    if (conversationFilter.includeTags.isNotEmpty) {
+      switch (conversationFilter.includeLogic) {
+        case FilterOperation.all:
+          if (!conversation.tagIds.toSet().containsAll(includeFilterTagIds)) return;
+          break;
+        case FilterOperation.any:
+          if (!conversation.tagIds.toSet().intersection(includeFilterTagIds).isNotEmpty) return;
+          break;
+      }
+    }
+
+    if (conversationFilter.excludeTags.isNotEmpty) {
+      switch (conversationFilter.excludeLogic) {
+        case FilterOperation.all:
+          if (conversation.tagIds.toSet().containsAll(excludeFilterTagIds)) return;
+          break;
+        case FilterOperation.any:
+          if (conversation.tagIds.toSet().intersection(excludeFilterTagIds).isNotEmpty) return;
+          break;
+      }
+    }
```

Comment:
Would it make sense moving this logic into a `ConversationFilter.filterConversation(conversation)` method?

Reply:
I agree, I think that makes sense. Done, thanks!